sklearn/metrics/pairwise.py

```diff
@@ -259,6 +297,137 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+def nan_euclidean_distances(X, Y=None, squared=False,
+                            missing_values=np.nan, copy=True):
+    """Calculate the euclidean distances in the presence of missing values.
+
+    Compute the euclidean distance between each pair of samples in X and Y,
+    where Y=X is assumed if Y=None. When calculating the distance between a
+    pair of samples, this formulation zero-weights feature coordinates with a
+    missing value in either sample and scales up the weight of the remaining
+    coordinates:
+
+        dist(x,y) = sqrt(weight * sq. distance from non-missing coordinates)
+        where,
+        weight = Total # of coordinates / # of non-missing coordinates
+
+    For example, the distance between ``[3, na, na, 6]`` and ``[1, na, 4, 5]``
+    is:
+
+        .. math::
+            \\sqrt{\\frac{4}{2}((3-1)^2 + (6-5)^2)}
+
+    If all the coordinates are missing or if there are no common non-missing
+    coordinates then NaN is returned for that pair.
+
+    Read more in the :ref:`User Guide <metrics>`.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples_1, n_features)
+
+    Y : {array-like, sparse matrix}, shape (n_samples_2, n_features)
+
+    squared : boolean, optional (default=False)
+        Return squared Euclidean distances.
+
+    missing_values : np.nan or integer, optional (default=np.nan)
```

Comment:
The `Imputer`class seems to accept the string `"NaN"` rather than `np.nan`. I would suggest staying consistent unless there's a good reason not to?

Reply:
I think we are moving away from using "NaN": https://github.com/scikit-learn/scikit-learn/pull/11211#issuecomment-395708181

`SimpleImputer` uses `np.nan` and `Imputer` is being depreciated.